In [18]:
import numpy as np
import pandas as pd
import pickle

from sklearn.model_selection import train_test_split
from sklearn.model_selection import GridSearchCV
from sklearn.preprocessing import MinMaxScaler
from sklearn.utils import resample
from sklearn.metrics import classification_report,accuracy_score

from sklearn.svm import SVC
from sklearn.multioutput import MultiOutputClassifier

from sklearn.linear_model import SGDClassifier
from sklearn.neighbors import KNeighborsClassifier

In [2]:
    with open('features.pkl', 'rb') as file:
        X = pickle.load(file)
        X=np.array(X)

    with open('labels.pkl', 'rb') as file:
        Y = pickle.load(file)
        Y=np.array(Y)

In [3]:
    scaler=MinMaxScaler()
    rescaled_data=scaler.fit_transform(X)
    X_train, X_test, y_train, y_test = train_test_split(rescaled_data,Y, test_size=0.2)

    y_test=np.array(y_test)
    y_train=np.array(y_train)

In [6]:
y_train.shape[0]+y_test.shape[0]

14825

In [ ]:
clf = MultiOutputClassifier(SVC(kernel='poly',class_weight='balanced',C=1, gamma=10))

#clf = MultiOutputClassifier(KNeighborsClassifier(n_neighbors=1))

clf.fit(X_train, y_train)

In [40]:
    #compute y_predicted
    y_pred = clf.predict(X_test)
    #compute accuracy
    accuracy = (y_pred == y_test).mean()
    print('Accuracy: {}'.format(accuracy))
    print('\n')
    #loop over the output columns to compute the f1 score, precision and recall
    columns=['y0','y1','y2','y3','y4','y5','y6','y7','y8','y9']
    y_pred=pd.DataFrame(y_pred,columns=columns)
    Y_test=pd.DataFrame(y_test,columns=columns)
    for i in columns: 
        print('Column: {}'.format(i))
        print(classification_report(Y_test[i],y_pred[i]))

Accuracy: 0.5477571669477235


Column: y0
              precision    recall  f1-score   support

           0       0.90      0.51      0.65      2503
           1       0.21      0.70      0.32       462

    accuracy                           0.54      2965
   macro avg       0.56      0.61      0.49      2965
weighted avg       0.79      0.54      0.60      2965

Column: y1
              precision    recall  f1-score   support

           0       0.89      0.54      0.68      2499
           1       0.21      0.65      0.32       466

    accuracy                           0.56      2965
   macro avg       0.55      0.60      0.50      2965
weighted avg       0.79      0.56      0.62      2965

Column: y2
              precision    recall  f1-score   support

           0       0.90      0.55      0.68      2609
           1       0.15      0.56      0.23       356

    accuracy                           0.55      2965
   macro avg       0.52      0.56      0.46      2965
weighted a

kernel=rbf c=1 gamma=1

Accuracy: 0.5237774030354132


Column: y0
              precision    recall  f1-score   support

           0       0.90      0.50      0.64      2503
           1       0.20      0.69      0.31       462

    accuracy                           0.53      2965
   macro avg       0.55      0.60      0.48      2965
weighted avg       0.79      0.53      0.59      2965

Column: y1
              precision    recall  f1-score   support

           0       0.90      0.52      0.66      2499
           1       0.21      0.68      0.32       466

    accuracy                           0.54      2965
   macro avg       0.55      0.60      0.49      2965
weighted avg       0.79      0.54      0.60      2965

Column: y2
              precision    recall  f1-score   support

           0       0.91      0.51      0.65      2609
           1       0.15      0.65      0.25       356

    accuracy                           0.53      2965
   macro avg       0.53      0.58      0.45      2965
weighted avg       0.82      0.53      0.61      2965

Column: y3
              precision    recall  f1-score   support

           0       0.86      0.45      0.59      2367
           1       0.24      0.70      0.36       598

    accuracy                           0.50      2965
   macro avg       0.55      0.57      0.47      2965
weighted avg       0.73      0.50      0.54      2965

Column: y4
              precision    recall  f1-score   support

           0       0.95      0.55      0.69      2711
           1       0.12      0.69      0.21       254

    accuracy                           0.56      2965
   macro avg       0.54      0.62      0.45      2965
weighted avg       0.88      0.56      0.65      2965

Column: y5
              precision    recall  f1-score   support

           0       0.80      0.38      0.52      2211
           1       0.29      0.73      0.41       754

    accuracy                           0.47      2965
   macro avg       0.55      0.55      0.46      2965
weighted avg       0.67      0.47      0.49      2965

Column: y6
              precision    recall  f1-score   support

           0       0.80      0.47      0.59      2216
           1       0.29      0.66      0.41       749

    accuracy                           0.52      2965
   macro avg       0.55      0.56      0.50      2965
weighted avg       0.67      0.52      0.54      2965

Column: y7
              precision    recall  f1-score   support

           0       0.92      0.55      0.69      2590
           1       0.17      0.65      0.27       375

    accuracy                           0.56      2965
   macro avg       0.54      0.60      0.48      2965
weighted avg       0.82      0.56      0.64      2965

Column: y8
              precision    recall  f1-score   support

           0       0.92      0.47      0.63      2604
           1       0.16      0.71      0.26       361

    accuracy                           0.50      2965
   macro avg       0.54      0.59      0.44      2965
weighted avg       0.83      0.50      0.58      2965

Column: y9
              precision    recall  f1-score   support

           0       0.86      0.51      0.64      2405
           1       0.23      0.64      0.34       560

    accuracy                           0.53      2965
   macro avg       0.55      0.57      0.49      2965
weighted avg       0.74      0.53      0.58      2965